specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,161 @@
+        // to this list.
+        LONG i = 0;
+        SAFEARRAY* list = SafeArrayCreateVector(VT_BSTR, 0, 1);
+        BSTR origin = SysAllocString(L"https://appassets.example");
+        SafeArrayPutElement(list, &i, origin);
+
+        VARIANT originList = {};
+        originList.parray = list;
+        originList.vt = VT_ARRAY;
+
+        // Add host object to a script for iframe access
+        CHECK_FAILURE(
+            webviewFrame->AddHostObjectToScriptWithOrigins(
+                L"sample", &remoteObjectAsVariant, &originList));
+
+        remoteObjectAsVariant.pdispVal->Release();
+        SysFreeString(origin);
+        SafeArrayDestroy(list);
+
+        return S_OK;
+    }).Get(), &m_frameCreatedToken));
+}
+```
+
+
+# Remarks
+List of origins in AddHostObjectToScript will be treated as following:
+ - empty list - no host object will be added for the iframe;
+ - list with origins - during access to host object from iframe the origin will be
+ checked that it belongs to this list;
+ - list with "*" element - host object will be available for iframe for all origins.
+ We suggest not to use this feature without understanding security implications of
+ giving access to host object from from iframes with unknown origins.
+
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+## IDL
+```c#
+/// WebView2Frame provides direct access to the iframes information and use.
+[uuid(18ae830b-8c83-459f-8a8c-95a1022af774), object, pointer_default(unique)]
+interface ICoreWebView2Frame : IUnknown {
+  /// The name of the iframe from the iframe html tag declaring it.
+  [propget] HRESULT Name([ out, retval ] LPWSTR * name);
+  /// Unique frame identifier.
+  [propget] HRESULT Id([ out, retval ] INT32* frameId);
+  /// Add the provided host object to script running in the WebView2 with the
+  /// specified name for the list of the specified origins. The host object
+  /// will be accessible for this iframe only if the iframe's origin during access
+  /// matches one of the origins which are passed.
+  /// \snippet ScenarioAddHostObject.cpp AddHostObjectToScriptWithOrigins
+  /// For more information about host objects navigate to [AddHostObjectToScript]
+  HRESULT AddHostObjectToScriptWithOrigins(
+      [in] LPCWSTR name,
+      [in] VARIANT *object,
+      [in] VARIANT *origins);
+}
+
+[uuid(ae64d3bc-01ba-4769-a470-a1888de6d30b), object, pointer_default(unique)]
+interface ICoreWebView2 : IUnknown {
+  /// Called when a new iframe is created. Use add_FrameDeleted to listen for
+  /// when this iframe goes away.
+  HRESULT add_FrameCreated(
+      [in] ICoreWebView2FrameEventHandler *eventHandler,
+      [out] EventRegistrationToken * token);
+  /// Remove an event handler previously added with add_FrameCreated.
+  HRESULT remove_FrameCreated([in] EventRegistrationToken token);
+
+  /// Called when iframe is deleted or its hosting process has died.
```

Comment:
It is interesting idea. So we will have frame created as WebView2 event and frame deleted as WebView2Frame?
Other that it is asymmetrical the user will need to subscribe to each frame deleted event separately instead of once, when it is WebView2 event.

Reply:
Maybe if you add sample code that demonstrates usage of the deleted event we can get a better idea of which would be better for the end developer. 

Subscribing to the deleted event on each frame individually is a benefit if the handler is a lambda capturing variables specific to that frame or otherwise if you want to perform different actions for different frames. With a deleted event on the CoreWebView2, the first thing the handler has to do is figure out which frame its being called for and then switching behavior based on that.

Reply:
Sample code for the frame deleted as WebView2 and WebView2Frame event.

1. FrameDeleted as WebView2 event:
```
CHECK_FAILURE(m_webview2->add_FrameDeleted(
        Callback<ICoreWebView2FrameEventHandler>(
            [this](
                ICoreWebView2* sender,
                ICoreWebView2FrameEventArgs* args) -> HRESULT {
        wil::com_ptr<ICoreWebView2Frame> webviewFrameDeleted;
        CHECK_FAILURE(args->get_Frame(&webviewFrameDeleted));
        LPWSTR name;
        webviewFrameDeleted->get_Name(&name);

        std::wstringstream message;
        message << "Frame " << std::wstring(name) << " will be deleted" << "\n";
        OutputDebugString(message.str().c_str());
        /*Cleanup*/

        return S_OK;
     }).Get(), &m_frameDeletedToken));
```

2. FrameDeleted as WebView2Frame event:

```
    CHECK_FAILURE(m_webView->add_FrameCreated(
        Callback<ICoreWebView2FrameEventHandler>(
            [this](
                ICoreWebView2* sender,
                ICoreWebView2FrameEventArgs* args) -> HRESULT
    {
        wil::com_ptr<ICoreWebView2Frame> webviewFrame;
        CHECK_FAILURE(args->get_Frame(&webviewFrame));

        webviewFrame->add_FrameDeleted(
        Callback<ICoreWebView2FrameDeletedEventHandler>(
            [this](ICoreWebView2* sender) -> HRESULT {
            /*Cleanup on frame deletion*/
        }).Get(), &m_frameDeletedToken));

        /*Add host object to script here*/
        
  }).Get(), &m_frameCreatedToken));

```
Please let me know which option do you prefer.

Reply:
The second looks easier to associate state between the frame and its deleted handle. Let's do that.